nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -43,11 +48,241 @@ use omicron_uuid_kinds::GenericUuid;
 use omicron_uuid_kinds::InstanceUuid;
 use omicron_uuid_kinds::PropolisUuid;
 use omicron_uuid_kinds::SledUuid;
+use slog::Logger;
+use std::collections::HashSet;
 use std::fmt;
 use strum::IntoEnumIterator;
 use thiserror::Error;
 use uuid::Uuid;
 
+#[derive(Debug, thiserror::Error)]
+enum SledReservationError {
+    #[error(
+        "Could not find any valid sled on which this instance can be placed"
+    )]
+    NotFound,
+    #[error("This instance belongs to an affinity group that requires it be placed \
+             on more than one sled. Instances can only placed on a single sled, so \
+             this is impossible to satisfy - Consider stopping other instances in \
+             the affinity group.")]
+    TooManyAffinityConstraints,
+    #[error(
+        "This instance belongs to an affinity group that requires it to be \
+             placed on a sled, but also belongs to an anti-affinity group that \
+             prevents it from being placed on that sled. These constraints are \
+             contradictory: Consider stopping instances in those \
+             affinity/anti-affinity groups, or changing group membership."
+    )]
+    ConflictingAntiAndAffinityConstraints,
+    #[error("This instance must be placed on a specific sled due to affinity \
+             rules, but that placement is invalid for some other reason (e.g., \
+             the instance would not fit, or the sled sharing an instance in the \
+             affinity group is being expunged). Consider stopping other \
+             instances in the affinity group, or changing affinity group \
+             membership.")]
+    RequiredAffinitySledNotValid,
```

Comment:
```suggestion
    #[error("This instance must be placed on a specific sled to co-locate it \
             with another instance in its affinity group, but that sled cannot \
             currently accept this instance. Consider stopping other instances \
             in the instance's affinity groups or changing its affinity group \
             membership.")]
    RequiredAffinitySledNotValid,
```

Optional language tweak: do we expose "expungement" as a concept to rack users who aren't operators? If not I might not mention it here; the suggested comment is hand-wavy about what it means for a sled to be "unable to accept an instance," but we could give examples here too (it has insufficient capacity or is being taken out of service).

I was going to suggest splitting these into separate "insufficient capacity" and "out of service" variants, but as you point out in the From impl below, this might require additional database queries/rummaging through returned rows to find out which case(s) apply.

Reply:
Sounds good, happy to update to this wording.

I was mulling around "how bad is this case, really" for the situation where sled expungement + affinity groups prevent allocation. I think it could only occur when:

1. A sled hosting a VMM is expunged, but the sled reservation still exists (hasn't been "failed" yet by the background tasks)
2. Another instance belonging to an affinity group with policy = fail is provisioned before that background task runs.

To actually mitigate this, I *think* we could look up the policy of all sleds returned by `lookup_(anti_)affinity_sleds_query`, and ignore all the sleds which are not provisionable.

However, I think the failure here should be brief (it'll fix itself with then failed VMM is stopped) and uncommon (it'll only happen when sharing an instance with a strict affinity policy, during expungement).